### PR TITLE
Add explicit derivative for jax.numpy.linalg.pinv.

### LINKED ISSUE #2792: Add support for differentiation of pseudoinverse
Consider the following example:
```python
 def f(p): 
    a = jnp.array([[0.,0.],[- p, 1.]]) * 1/ (1 + p**2) 
    return jnp.linalg.pinv(a) 


jax.jacobian(f)(2.)                                                                                                                                 
Out[75]: 
DeviceArray([[nan, nan],
             [nan, nan]], dtype=float32)
```
So it returns `nan` values, whereas the correct values are
```
array(
[[0., -1.], 
[ 0., 0.]])
```
Oddly enough, jax is able to handle this function correctly:
```python
def f2(p): 
   a = jnp.array([[- p, 1.]]) * 1/ (1 + p**2) 
return jnp.linalg.pinv(a)                                                                                                                        

In [77]: jax.jacobian(f)(2.)                                                                                                                                 
Out[77]: 
DeviceArray([[-0.99999976],
             [ 0.        ]], dtype=float32)
```

Thoughts:
1. ~~f2 returns nd arrays and jax.jacobian seems not to handle those cases to well (probably should open another issue for that)~~ (this seems to be fixed in recent versions though) 

2. There is a closed formula for the derivative of the pseudoinverse which uses the derivative of the original array: https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse#Derivative

### PR DESCRIPTION
**Pull request by hawkinsp**
Fix type confusion problems in the JVP rule for SVD that meant it produced 64-bit tangents for 32-bit primals.

Fixes #2792

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
LGTM, I suspect there's a better way to associate things in the formula to improve numerics, though I am not expert enough to offer concrete suggestions. Maybe Rasmus could help.

**Comment by mattjj** [comment]
The first term is the regular matrix inverse derivative formula for invertible matrices, and the latter two terms are zero when A^+ = A^{-1}.

The first term is of the form X -> A X A and the latter two terms are of the form X -> B X C' + C X B'. I wonder if there are numerical issues that arise just from how we associate those matrix products. Or whether we expand C = (I - E) and distribute into X -> B X - B X E' + X B' - E X B'.

Anyway just wanted to offer some raw thoughts, maybe those ideas are not useful!

**Comment by mattjj** [comment]
Or maybe it's just fundamentally unstable because of how pseudoinverses work...

**Comment by mattjj** [comment]
It may be that the most numerically stable thing to do would reuse the SVD decomposition of the input matrix, since then we can e.g. treat the numerically nonzero singular values and their corresponding subspaces differently from the others.

**Comment by hawkinsp** [comment]
Are you happy merging this as is?

(I note this is the same rule as autograd, FWIW.)

File: jax/numpy/linalg.py [hunk 1]
```diff
@@ -188,32 +189,45 @@ def eigvalsh(a, UPLO='L'):
...
+def _pinv_jvp(rcond, primals, tangents):
+  # The Differentiation of Pseudo-Inverses and Nonlinear Least Squares Problems
+  # Whose Variables Separate. Author(s): G. H. Golub and V. Pereyra. SIAM
+  # Journal on Numerical Analysis, Vol. 10, No. 2 (Apr., 1973), pp. 413-432.
+  # (via https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse#Derivative)
+  a, = primals
+  a_dot, = tangents
+  p = pinv(a, rcond=rcond)
+  m, n = a.shape[-2:]
+  p_dot = -p @ a_dot @ p
```
Comment (shoyer) [inline jax/numpy/linalg.py:225]: I guess TPUs don't have SVD yet so it's OK to use `@`, but at least let's add a comment noting that this we would need higher precision matmul for TPUs.

File: jax/numpy/linalg.py [hunk 2]
```diff
@@ -188,32 +189,45 @@ def eigvalsh(a, UPLO='L'):
...
+  p_dot = p_dot + p @ _H(p) @ _H(a_dot) @ (np.eye(m, dtype=a.dtype) - a @ p)
+  p_dot = p_dot + (np.eye(n, dtype=a.dtype) - p @ a) @ _H(a_dot) @ _H(p) @ p
```
Comment (shoyer) [inline jax/numpy/linalg.py:227]: I suspect this could be simplified somewhat (in the Hermitian case) if we added support for `hermitian=True` from NumPy.

**Comment by mattjj** [comment]
Yes, very happy!

File: tests/linalg_test.py [hunk 3]
```diff
@@ -716,6 +716,17 @@ def testPinv(self, shape, dtype, rng_factory):
     self._CheckAgainstNumpy(onp.linalg.pinv, np.linalg.pinv, args_maker,
                             check_dtypes=True, tol=1e-3)
     self._CompileAndCheck(np.linalg.pinv, args_maker, check_dtypes=True)
+    # TODO(phawkins): 1e-1 seems like a very loose tolerance.
+    jtu.check_grads(np.linalg.pinv, args_maker(), 2, rtol=1e-1)
+
+
+  def testPinvGradIssue2792(self):
+    def f(p):
+      a = np.array([[0., 0.],[-p, 1.]], np.float32) * 1 / (1 + p**2)
```
Comment (shoyer) [inline tests/linalg_test.py:725]: Another scenario where I think the current gradient implementation would fail is any case where there are repeated singular values in the SVD, e.g., `np.eye(2)`.

**Review by shoyer — APPROVED** [review]

**Comment by hawkinsp** [comment]
Actually I note that TF has the following formulation of the cutoff which I think also fixes the NaN problem:
```
  s = np.where(s > cutoff, s, np.inf)
  res = np.matmul(_T(v), np.divide(_T(u), s[..., np.newaxis]))
```
(instead of:
```
  s = np.where(s > cutoff, np.divide(1, s), 0)
  res = np.matmul(_T(v), np.multiply(s[..., np.newaxis], _T(u)))
```
)

Maybe that's simpler anyway and we don't need a custom JVP at all?

Interestingly it seems to fail only for some complex cases, which is odd.

**Comment by mattjj** [comment]
Nice find!

**Comment by shoyer** [comment]
Here's another fail case with the current algorithm:
```
>>> jax.jacobian(jnp.linalg.pinv)(jnp.eye(2))
DeviceArray([[[[nan, nan],
               [nan, nan]],

              [[nan, nan],
               [nan, nan]]],


             [[[nan, nan],
               [nan, nan]],

              [[nan, nan],
               [nan, nan]]]], dtype=float32)
```

**Comment by mattjj** [comment]
Hm and that matrix looks invertible :)

File: jax/numpy/linalg.py (hunk 1 above)
Comment (hawkinsp) [reply jax/numpy/linalg.py:226]: Done.

File: tests/linalg_test.py (hunk 3 above)
Comment (hawkinsp) [reply tests/linalg_test.py:725]: Added `np.eye(2)` as a test case. This passes under the new gradient but failed under both variants of differentiating through the SVD.

File: jax/numpy/linalg.py (hunk 2 above)
Comment (hawkinsp) [reply jax/numpy/linalg.py:229]: Added a TODO about the Hermitian case.

**Review by shoyer — APPROVED** [review]

**Comment by dbankmann** [comment]
> The first term is the regular matrix inverse derivative formula for invertible matrices, and the latter two terms are zero when A^+ = A^{-1}.
> 
> The first term is of the form X -> A X A and the latter two terms are of the form X -> B X C' + C X B'.
Just a few more thoughts:
A^+A  is not the transpose of AA^+, those are orthogonal projectors and thus self-adjoint.

The derivatives of A^+A  or AA^+ will zero out the second or last term  of d A^+, respectively, which may not be the case for a badly conditioned A.

**Comment by mattjj** [comment]
#2744 may be related, in that instead of relying on the SVD derivative rule we might want to have a custom JVP rule that uses the results of the SVD (and in particular works more directly on the numerically identified row space / col space / null space / col space perp).

**Comment by shoyer** [comment]
> #2744 may be related, in that instead of relying on the SVD derivative rule we might want to have a custom JVP rule that uses the results of the SVD (and in particular works more directly on the numerically identified row space / col space / null space / col space perp).

Indeed, practically speaking I suspect the right gradient rule to use in both cases is probably `lax.custom_linear_solve`. Mathematically the implicit function theorem math works the same for either exact solves or psuedo-inverse solves or least-square solves.

The interface would work for `lstsq` (I think) but unfortunately not for `pinv`. The problem is that we really need a "solve" interface, i.e.,
```
def psolve(a, b):
  return jnp.linalg.pinv(a) @ b
```
and it appears neither numpy nor scipy have a `psolve` function

**Comment by shoyer** [comment]
Wow -- using custom_linear_solve is 10x faster (and also numerically stable): https://github.com/google/jax/issues/2802

We still need this custom_jvp for `pinv`, but most users probably shouldn't be differentiating `pinv` directly unless they're using it for something other than solves.
